workflow-ui/backstack-android/src/main/java/com/squareup/workflow1/ui/backstack/StateRegistryCache.kt

```diff
@@ -0,0 +1,211 @@
+        parentRegistryOwner!!.savedStateRegistry.consumeRestoredStateForKey(parentKey!!)
+      performRestore(restoredState)
+    }
+  }
+
+  private val stateProvider = SavedStateProvider {
+    Bundle().apply {
+      doIfRestored { states ->
+        // Give the current view a chance to save itself.
+        currentOwner?.let { owner ->
+          val currentState = Bundle()
+          owner.controller.performSave(currentState)
+          states += owner.key to currentState
+        }
+
+        // Convert states map to a bundle.
+        states.forEach { (key, state) -> putBundle(key, state) }
+      }
+    }
+  }
+
+  public fun attach(
+    key: String,
+    parentOwner: SavedStateRegistryOwner
+  ) {
+    check(parentRegistryOwner == null) {
+      "Expected StateRegistryCache to be detached before re-attaching. " +
+        "Already attached to $parentRegistryOwner but trying to attach to $parentOwner"
+    }
+    this.parentRegistryOwner = parentOwner
+    this.parentKey = key
+
+    // We can only be restored once.
+    if (isRestored) return
+
+    val parentRegistry = parentOwner.savedStateRegistry
+    val parentLifecycle = parentOwner.lifecycle
+
+    // If the key is already registered, SavedStateRegistry throws an exception that doesn't provide
+    // enough information to troubleshoot, so we add some ourselves.
+    try {
+      parentRegistry.registerSavedStateProvider(key, stateProvider)
+    } catch (e: IllegalArgumentException) {
+      throw IllegalArgumentException(
+        "Error registering StateRegistryCache as SavedStateProvider with key \"$key\" on parent " +
+          "SavedStateRegistryOwner $parentOwner.\n"+
+        "You might need to set a view ID on your BackStackContainer or wrap your BackStackScreen " +
+          "with a Named rendering.",
+        e
+      )
+    }
+
+    if (parentRegistry.isRestored) {
+      val restoredState = parentRegistry.consumeRestoredStateForKey(key)
+      performRestore(restoredState)
+    } else {
+      val currentState = parentLifecycle.currentState
+      check(currentState != INITIALIZED) {
+        "Expected parent lifecycle to be in state INITIALIZED, but was $currentState"
+      }
+
+      // Parent registry hasn't been restored yet, so wait for the signal that it will be ready.
+      parentLifecycle.addObserver(lifecycleObserver)
+    }
+  }
+
+  public fun detach() {
+    parentRegistryOwner!!.savedStateRegistry.unregisterSavedStateProvider(parentKey!!)
+    parentRegistryOwner!!.lifecycle.removeObserver(lifecycleObserver)
+    parentRegistryOwner = null
+  }
+
+  /**
+   * To be called when the set of hidden views changes but the visible view remains
+   * the same. Any cached view state held for renderings that are not
+   * [compatible][com.squareup.workflow1.ui.compatible] those in [retaining] will be dropped.
+   */
+  public fun prune(retaining: Collection<Named<*>>) {
+    pruneKeys(retaining.map { it.compatibilityKey })
+  }
+
+  public fun update(
```

Comment:
- [ ] This duplicates a lot of non-trivial logic in `ViewStateCache`. Figure out how to de-dup that. Maybe the interesting bits of this class, the state machine for coordinating registries, can be pulled out into a helper, and the rest of the work just merged into `ViewStateCache`?